File: lib/py/uapi/internal/validation/ValidateSelect.py
from typing import TYPE_CHECKING, cast

if TYPE_CHECKING:
    from uapi.internal.validation.ValidateContext import ValidateContext
    from uapi.internal.validation.ValidationFailure import ValidationFailure


def validate_select(given_obj: object, possible_fn_selects: dict[str, object], ctx: 'ValidateContext') -> list['ValidationFailure']:
    from uapi.internal.validation.GetTypeUnexpectedValidationFailure import get_type_unexpected_validation_failure

    if not isinstance(given_obj, dict):
        return get_type_unexpected_validation_failure([], given_obj, 'Object')

    fn_scope = cast(str, ctx.fn)

    possible_select = possible_fn_selects[fn_scope]

    return is_sub_select([], given_obj, possible_select)


def is_sub_select(path: list[object], given_obj: object, possible_select_section: object) -> list['ValidationFailure']:
    from uapi.internal.validation.GetTypeUnexpectedValidationFailure import get_type_unexpected_validation_failure
    from uapi.internal.validation.ValidationFailure import ValidationFailure

    if isinstance(possible_select_section, list):
        if not isinstance(given_obj, list):
            return get_type_unexpected_validation_failure(path, given_obj, 'Array')

        validation_failures = []

        for index, element in enumerate(given_obj):
            if element not in possible_select_section:
                validation_failures.append(ValidationFailure(
                    path + [index], 'ArrayElementDisallowed', {}))

        return validation_failures

    elif isinstance(possible_select_section, dict):
        if not isinstance(given_obj, dict):
            return get_type_unexpected_validation_failure(path, given_obj, 'Object')

        validation_failures = []

        for key, value in given_obj.items():
            if key in possible_select_section:
                inner_failures = is_sub_select(
                    path + [key], value, possible_select_section[key])
                validation_failures.extend(inner_failures)
            else:
                validation_failures.append(ValidationFailure(
                    path + [key], 'ObjectKeyDisallowed', {}))

        return validation_failures

    return []

File: lib/java/src/main/java/uapi/internal/validation/ValidateSelect.java
package uapi.internal.validation;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ValidateSelect {

    public static List<ValidationFailure> validateSelect(Object givenObj,
            Map<String, Object> possibleFnSelects, ValidateContext ctx) {
        if (!(givenObj instanceof Map) || givenObj == null) {
            return GetTypeUnexpectedValidationFailure.getTypeUnexpectedValidationFailure(new ArrayList<>(), givenObj,
                    "Object");
        }

        Map<String, Object> possibleSelect = (Map<String, Object>) possibleFnSelects.get(ctx.fn);

        return isSubSelect(new ArrayList<>(), givenObj, possibleSelect);
    }

    private static List<ValidationFailure> isSubSelect(List<Object> path, Object givenObj,
            Object possibleSelectSection) {

        if (possibleSelectSection instanceof List) {
            if (!(givenObj instanceof List)) {
                return GetTypeUnexpectedValidationFailure.getTypeUnexpectedValidationFailure(path, givenObj, "Array");
            }

            List<ValidationFailure> validationFailures = new ArrayList<>();

            List<?> givenObjList = (List<?>) givenObj;
            List<?> possibleSelectList = (List<?>) possibleSelectSection;

            for (int index = 0; index < givenObjList.size(); index++) {
                Object element = givenObjList.get(index);
                if (!possibleSelectList.contains(element)) {
                    var loopPath = new ArrayList<>(path);
                    loopPath.add(index);
                    validationFailures.add(new ValidationFailure(loopPath, "ArrayElementDisallowed", Map.of()));
                }
            }

            return validationFailures;
        } else if (possibleSelectSection instanceof Map) {
            if (!(givenObj instanceof Map) || givenObj == null) {
                return GetTypeUnexpectedValidationFailure.getTypeUnexpectedValidationFailure(path, givenObj, "Object");
            }

            List<ValidationFailure> validationFailures = new ArrayList<>();

            Map<String, Object> givenObjMap = (Map<String, Object>) givenObj;
            Map<String, Object> possibleSelectMap = (Map<String, Object>) possibleSelectSection;

            for (Map.Entry<String, Object> entry : givenObjMap.entrySet()) {
                String key = entry.getKey();
                Object value = entry.getValue();
                if (possibleSelectMap.containsKey(key)) {
                    var loopPath = new ArrayList<>(path);
                    loopPath.add(key);
                    List<ValidationFailure> innerFailures = isSubSelect(loopPath, value, possibleSelectMap.get(key));
                    validationFailures.addAll(innerFailures);
                } else {
                    validationFailures.add(new ValidationFailure(new ArrayList<>(path) {
                        {
                            add(key);
                        }
                    }, "ObjectKeyDisallowed", Map.of()));
                }
            }

            return validationFailures;
        }

        return new ArrayList<>();
    }
}
